lib/puppet/provider/ec2_instance/v2.rb

```diff
@@ -105,13 +130,57 @@ def create
       key = resource[:key_name] ? resource[:key_name] : false
       config['key_name'] = key if key
 
-      response = ec2_client(resource[:region]).run_instances(config)
+      # must specify a subnet to launch inside a VPC
+      if ! resource[:subnet]
+        matching_groups = classic_groups
+      else
+        # filter by VPC, since describe_subnets doesn't work on empty tag:Name
+        subnet_response = ec2.describe_subnets(filters: [
+          {name: "vpc-id", values: vpc_groups.keys}])
+
+        # then find the name in the VPC subnets that we have
+        subnets = subnet_response.data.subnets.select do |s|
+          if resource[:subnet].empty?
+            ! s.tags.any? { |t| t.key == 'Name' }
+          else
+            s.tags.any? { |t| t.key == 'Name' && t.value == resource[:subnet] }
+          end
+        end
+
+        # handle ambiguous name collisions by selecting first matching subnet / vpc
+        subnet = subnets.first
+        if subnets.length > 1
+          subnet_map = subnets.map { |s| "#{s.subnet_id} (vpc: #{s.vpc_id})" }.join(', ')
+          Puppet.warning "Ambiguous subnet name '#{resource[:subnet]}' resolves to subnets #{subnet_map} - using #{subnet.subnet_id}"
+        end
+
+        if subnet.nil?
+          raise Puppet::Error,
+            "Invalid security groups '#{groups.join(', ')}' not found in VPCs '#{vpc_groups.keys.join(', ')}'"
```

Comment:
This would be more friendly if it raised an error earlier than this if vpc_groups is empty. Currently, it claims the group was "`not found in VPCs ''`" which is hard to read.